File: DiKErnel/Src/DiKErnel.Integration/Internal/Helpers/ValidationHelper.cpp
// Copyright (C) Stichting Deltares 2020. All rights reserved.
//
// This file is part of DiKErnel.
//
// DiKErnel is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, version 3.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.
//
// All names, logos, and references to "Deltares" are registered trademarks of
// Stichting Deltares and remain full property of Stichting Deltares at all times.
// All rights reserved.

#include "ValidationHelper.h"

#include <stdexcept>

#include "EventRegistry.h"

namespace DiKErnel::Integration
{
    using namespace DomainLibrary;
    using namespace std;
    using namespace Util;

    bool ValidationHelper::RegisterValidationIssues(
        const vector<unique_ptr<ValidationIssue>>& validationIssues)
    {
        auto validationSuccessful = true;

        for (auto& validationIssue : validationIssues)
        {
            auto validationForIssueSuccessful = true;

            if (validationIssue != nullptr)
            {
                validationForIssueSuccessful = RegisterValidationIssue(*validationIssue);
            }

            validationSuccessful = validationForIssueSuccessful && validationSuccessful;
        }

        return validationSuccessful;
    }

    bool ValidationHelper::RegisterValidationIssue(
        const ValidationIssue& validationIssue)
    {
        const auto validationIssueType = validationIssue.GetValidationIssueType();

        EventRegistry::Register(make_unique<Event>(validationIssue.GetMessage(),
                                                   ConvertValidationIssueType(validationIssueType)));

        return validationIssueType != ValidationIssueType::Error;
    }

    EventType ValidationHelper::ConvertValidationIssueType(
        const ValidationIssueType validationIssueType)
    {
        switch (validationIssueType)
        {
            case ValidationIssueType::Warning:
                return EventType::Warning;
            case ValidationIssueType::Error:
                return EventType::Error;
            default:
                throw out_of_range("Invalid ValidationIssueType.");
        }
    }
}

File: Src/DiKErnel.Integration/Helpers/ValidationHelper.cs
// Copyright (C) Stichting Deltares and State of the Netherlands 2025. All rights reserved.
//
// This file is part of DiKErnel.
//
// This program is free software; you can redistribute it and/or modify it under the terms of
// the GNU Lesser General Public License as published by the Free Software Foundation; either
// version 3 of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
// without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
// See the GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License along with this
// program; if not, see <https://www.gnu.org/licenses/>.
//
// All names, logos, and references to "Deltares" are registered trademarks of Stichting
// Deltares and remain full property of Stichting Deltares at all times. All rights reserved.

using System;
using System.Collections.Generic;
using DiKErnel.Util;
using DiKErnel.Util.Validation;

namespace DiKErnel.Integration.Helpers
{
    /// <summary>
    /// Helper for registering validation issues.
    /// </summary>
    public static class ValidationHelper
    {
        /// <summary>
        /// Registers the provided validation issues.
        /// </summary>
        /// <param name="validationIssues">The validation issues to register.</param>
        /// <returns><c>false</c> when one or more validation issues are of an error type;
        /// <c>true</c> otherwise.</returns>
        public static bool RegisterValidationIssues(IReadOnlyList<ValidationIssue> validationIssues)
        {
            var overallValidationSuccessful = true;

            foreach (ValidationIssue validationIssue in validationIssues)
            {
                overallValidationSuccessful &= validationIssue == null || RegisterValidationIssue(validationIssue);
            }

            return overallValidationSuccessful;
        }

        private static bool RegisterValidationIssue(ValidationIssue validationIssue)
        {
            ValidationIssueType validationIssueType = validationIssue.ValidationIssueType;

            EventRegistry.Register(new Event(validationIssue.Message, ConvertValidationIssueType(validationIssueType)));

            return validationIssueType != ValidationIssueType.Error;
        }

        private static EventType ConvertValidationIssueType(ValidationIssueType validationIssueType)
        {
            return validationIssueType switch
            {
                ValidationIssueType.Warning => EventType.Warning,
                ValidationIssueType.Error => EventType.Error,
                _ => throw new NotSupportedException("Unsupported ValidationIssueType.")
            };
        }
    }
}